### PR TITLE
Add concepts for transfusions

### PR DESCRIPTION
**Pull request by nsfinkelstein**
I think all relevant itemids are included but I could be mistaken.

The query supposes that if any transfusion is started within 1 hr of the last, it should be considered part of the same 'transfusion event' and the amounts of all transfusions so connected should be summed.

Thanks
Noam

### CODE REVIEW

**Comment by n-s-f** [comment]
I've changed the logic to group by `linkorderid` instead of assuming that all starts within 1hr apart are part of the same event. The results are similar.

**Comment by n-s-f** [comment]
@alfredoncs 

`tsp` is short for timestamp. I use it because the names of the starting timestamp columns in CV and MV are different from each other (`charttime` and `starttime`), and they have to be the same to be unioned.

I group by `linkorderid` because according to the [documentation](https://mimic.physionet.org/mimictables/inputevents_mv/#orderid-linkorderid) all records that are related to the same order will have the same `linkorderid`. So for example of the order is for a 1000ml RBC transfusion, that might show up as two separate 500ml transfusions, with one record for each half liter bag. This view supposes that it's more useful to know that 1000ml were transfused under the same order than to care about how those 1000ml were split between bags or what the rate changes were.

**Comment by n-s-f** [comment]
@alfredoncs 

`tsp` is short for timestamp. I use it because the names of the starting timestamp columns in CV and MV are different from each other (`charttime` and `starttime`), and they have to be the same to be unioned.

I group by `linkorderid` because according to the [documentation](https://mimic.physionet.org/mimictables/inputevents_mv/#orderid-linkorderid) all records that are related to the same order will have the same `linkorderid`. So for example of the order is for a 1000ml RBC transfusion, that might show up as two separate 500ml transfusions, with one record for each half liter bag. This view supposes that it's more useful to know that 1000ml were transfused under the same order than to care about how those 1000ml were split between bags or what the rate changes were.

**Comment by alfredoncs** [comment]
Ok.
Thank you.
Best

Alfredo N C Santana, MD, PhD, FCCP
Professor Permanente - Mestrado ESCS
Professor Adjunto - Medicina ESCS
Pneumologista - HRAN SES DF

Em Seg, 2 de abr de 2018 12:25, Noam Finkelstein <notifications@github.com>
escreveu:

**Comment by alistairewj** [comment]
Yeah I never considered using linkorderid but it makes sense. I think it's worth some investigation to see if this works on CareVue though. I frequently recall, for other drugs, multiple administrations being classified under the same ICU system order because it was convenient, and not necessarily because they were at the same time. Perhaps a distribution of `charttime - min(charttime)` grouped by `linkorderid`?

Also, why do you elect to include PACU and pre-ICU transfusions? Obviously it's useful information - but in these cases, the charttime is not at all accurate and they are only documented to indicate that  a patient has had these transfusions. If you include them, you end up with huge spikes of 7-8 units "on admission" when in reality the patient just came from the OR. Wondering what your thoughts are on this.

**Comment by n-s-f** [comment]
@alistairewj Thanks for the comments. I misunderstood the recording of PACU and pre-ICU charttimes - I'll remove those itemids. Are the OR charttimes valid (e.g. itemid 30103 (OR FFP) in CareVue)?

I ran the following query to see the distribution of timespans covered by the same linkorderid.

```.sql
with bins as (
  select array_agg(distinct itemid) as itemid
    , width_bucket(EXTRACT(epoch FROM max(charttime) - min(charttime))/3600, 0, 480, 48) as bin
  from inputevents_cv                                                                                                                                  
  where itemid in (                                                                                                                                    
      30005,  -- Fresh Frozen Plasma                                                                                                                   
      30180,  -- Fresh Froz Plasma                                                                                                                     
      30103,  -- OR FFP                                                                                                                                
      44172,  -- FFP GTT                                                                                                                               
      44236,  -- E.R. FFP                                                                                                                              
      46410,  -- angio FFP                                                                                                                             
      46418,  -- ER ffp                                                                                                                                
      46684,  -- ER FFP                                                                                                                                
      44819,  -- FFP ON FARR 2                                                                                                                         
      46530,  -- Floor FFP                                                                                                                             
      44044,  -- FFP Drip                                                                                                                              
      46122,  -- ER in FFP                                                                                                                             
      45669,  -- ED FFP                                                                                                                                
      42323   -- er ffp                                                                                                                                
    )                                                                                                                                                  
    and amount > 0                                                                                                                                     
  group by linkorderid
)

select ((bin - 1) * 10)::text || ' to ' || (bin*10)::text as hours, count(*), itemid
from bins group by itemid, bin
order by itemid, bin    
```

This results in: 

```
+------------+---------+-------------------------------------------------------------------------------+
| hours      |   count | itemid                                                                        |
|------------+---------+-------------------------------------------------------------------------------|
| 0 to 10    |   2,847 | [30005]                                                                       |
| 10 to 20   |     275 | [30005]                                                                       |
| 20 to 30   |     196 | [30005]                                                                       |
| 30 to 40   |     116 | [30005]                                                                       |
| 40 to 50   |      65 | [30005]                                                                       |
| 50 to 60   |      52 | [30005]                                                                       |
| 60 to 70   |      49 | [30005]                                                                       |
| 70 to 80   |      48 | [30005]                                                                       |
| 80 to 90   |      28 | [30005]                                                                       |
| 90 to 100  |      20 | [30005]                                                                       |
| 100 to 110 |      14 | [30005]                                                                       |
| 110 to 120 |      20 | [30005]                                                                       |
| 120 to 130 |      16 | [30005]                                                                       |
| 130 to 140 |      15 | [30005]                                                                       |
| 140 to 150 |      10 | [30005]                                                                       |
| 150 to 160 |      10 | [30005]                                                                       |
| 160 to 170 |      11 | [30005]                                                                       |
| 170 to 180 |       5 | [30005]                                                                       |
| 180 to 190 |       6 | [30005]                                                                       |
| 190 to 200 |       7 | [30005]                                                                       |
| 200 to 210 |       2 | [30005]                                                                       |
| 210 to 220 |       4 | [30005]                                                                       |
| 220 to 230 |       7 | [30005]                                                                       |
| 230 to 240 |       4 | [30005]                                                                       |
| 240 to 250 |       4 | [30005]                                                                       |
| 250 to 260 |       1 | [30005]                                                                       |
| 260 to 270 |       6 | [30005]                                                                       |
| 270 to 280 |       2 | [30005]                                                                       |
| 280 to 290 |       2 | [30005]                                                                       |
| 290 to 300 |       2 | [30005]                                                                       |
| 310 to 320 |       4 | [30005]                                                                       |
| 320 to 330 |       2 | [30005]                                                                       |
| 330 to 340 |       4 | [30005]                                                                       |
| 340 to 350 |       1 | [30005]                                                                       |
| 360 to 370 |       4 | [30005]                                                                       |
| 370 to 380 |       4 | [30005]                                                                       |
| 380 to 390 |       5 | [30005]                                                                       |
| 390 to 400 |       1 | [30005]                                                                       |
| 420 to 430 |       2 | [30005]                                                                       |
| 430 to 440 |       3 | [30005]                                                                       |
| 440 to 450 |       1 | [30005]                                                                       |
| 450 to 460 |       1 | [30005]                                                                       |
| 470 to 480 |       2 | [30005]                                                                       |
| 480 to 490 |      36 | [30005]                                                                       |
| 0 to 10    |     835 | [30103]                                                                       |
| 10 to 20   |      13 | [30103]                                                                       |
| 20 to 30   |       5 | [30103]                                                                       |
| 30 to 40   |       5 | [30103]                                                                       |
| 40 to 50   |       1 | [30103]                                                                       |
| 50 to 60   |       1 | [30103]                                                                       |
| 90 to 100  |       2 | [30103]                                                                       |
| 0 to 10    |      21 | [30180]                                                                       |
| 10 to 20   |       1 | [30180]                                                                       |
| 20 to 30   |       2 | [30180]                                                                       |
| 30 to 40   |       1 | [30180]                                                                       |
| 60 to 70   |       1 | [30180]                                                                       |
| 70 to 80   |       1 | [30180]                                                                       |
| 80 to 90   |       1 | [30180]                                                                       |
| 100 to 110 |       1 | [30180]                                                                       |
| 120 to 130 |       1 | [30180]                                                                       |
| 150 to 160 |       1 | [30180]                                                                       |
| 200 to 210 |       1 | [30180]                                                                       |
| 210 to 220 |       1 | [30180]                                                                       |
| 480 to 490 |       1 | [42323, 44044, 44172, 44236, 44819, 45669, 46122, 46410, 46418, 46530, 46684] |
+------------+---------+-------------------------------------------------------------------------------+
SELECT 64
```

It seems you are right that linkedorderids regularly span more time than what should be considered a single transfusion.

My previous approach was to consider a string of records to be part of the same transfusion as long as no two consecutive records in that string are separated by more than an hour. The query in that case looks like this: https://github.com/n-s-f/mimic-code/blob/5d0e6cdbb1e1fc712156c6e08bd1e0aa0c00cc76/concepts/cookbook/rbc_transfusion.sql

Does that seem reasonable to you?

Thanks

File: concepts/cookbook/rbc_transfusion.sql [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      30179,  -- PRBC's
+      42324,  -- er prbc
+      42588,  -- VICU PRBC
+      30001,  -- Packed RBC's
+      30004,  -- Washed PRBC's
+      42239,  -- CC7 PRBC
+      46407,  -- ED PRBC
+      46612,  -- E.R. prbc
+      46124,  -- er in prbc
+      42740   -- prbc in er
```
Comment (alistairewj) [inline concepts/cookbook/rbc_transfusion.sql]: this line and the 4-5 above it all seem to be emergency room transfusions

File: concepts/cookbook/ffp_transfusion.sql [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      44236,  -- E.R. FFP        
+      46410,  -- angio FFP
+      46418,  -- ER ffp
+      46684,  -- ER FFP
+      44819,  -- FFP ON FARR 2
+      46530,  -- Floor FFP       
+      44044,  -- FFP Drip
+      46122,  -- ER in FFP
+      45669,  -- ED FFP
+      42323   -- er ffp
```
Comment (alistairewj) [inline concepts/cookbook/ffp_transfusion.sql]: this line and the 3 above it all seem to be emergency room transfusions

File: concepts/cookbook/ffp_transfusion.sql [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  where itemid in (
+      30005,  -- Fresh Frozen Plasma
+      30180,  -- Fresh Froz Plasma
+      44172,  -- FFP GTT         
+      44236,  -- E.R. FFP        
+      46410,  -- angio FFP
+      46418,  -- ER ffp
+      46684,  -- ER FFP
+      44819,  -- FFP ON FARR 2
+      46530,  -- Floor FFP
```
Comment (alistairewj) [inline concepts/cookbook/ffp_transfusion.sql]: this line and the 6 above it all seem to be emergency room or floor transfusions

**Comment by nsfinkelstein** [comment]
This pull request is superseded by https://github.com/MIT-LCP/mimic-code/pull/714.

**Pull request closed without merging**